### ISSUE TITLE
amp-ima-video: Use native controls.

### ISSUE THREAD

**Issue by aghassemi** [ampproject/amphtml#8841]
With using native controls we will get UX consistency, built-in accessibility,Volume, Chromecast, AirPlay, Closed Caption, etc.. 

There are issues with using native controls in IMA, that need addressing however:

1- Initially native controls need to be hidden and a custom "play" icon for click to play need to be present to capcture user-intent and initialize IMA.
2- During Ad play, controls need to be hidden since there is no way to disable "seek" otherwise.
3- Native controls `Fullscreen` is a big issue, outside of iOS, IMA needs to have a different Ad container rather than the `<video>` itself to go to fullscreen so various Ad formats can be served. There are two approaches for this:
   - In Chrome, we can hide the native fullscreen with `video::-webkit-media-controls-fullscreen-button { display: none }` and provide our own button.
   - In FF/IE, since we are in an iframe, we can toggle iframe's `allowfullscreen` attribute which automatically removes the native fullscreen control from all `<video>`s inside the iframe.

**Comment by shawnbuso** [comment]
I'm not sure if this one is technically feasible. So far I have exactly what we need working great on desktop and Android, but iPhone is giving me all kinds of problems. First, it won't let me disable controls during ads - I'm calling `videoPlayer.setAttribute('controls', false)`, but it seems to just be ignored - I still see controls while ads are playing, which interferes with our ad UI. \

Second, and this is the bigger issue, I cannot get the fullscreen button to work. I have a fullscreen button overlayed on my player, and for some reason I can tap this once and only once. If I enter fullscreen then exit fullscreen, I cannot re-enter fullscreen. Either something is interfering with my ability to tap the button (which should be the case because I have z-index set to max int) or my event listener is being removed, ignored, or interrupted. This works fine on desktop and Android - it's only iPhone that's giving me problems. Have you seen anything remotely like this before (something that can be tapped once but future taps are ignored)? I'm not getting any errors (related to this) in my console.

You can see my work in progress at https://github.com/shawnbuso/amphtml/tree/ima_controls.

**Comment by shawnbuso** [comment]
OK after a full weekend of shower-levels of thinking I think I've resolved both of these. For anyone else stumbling upon this with a similar issue:

For the first issue, I had to change `videoPlayer.setAttribute('controls', false);` to `videoPlayer.removeAttribute('controls');` It looks like maybe iPhone just looks for the presence of the controls attribute and disregards the value associated with it.

For the second issue, this was just an error in my logic - when I exited fullscreen, I forgot to flip the bit that said I was back in regular playback mode, so subsequent attempts to enter fullscreen just re-ran the exit fullscreen logic. Also my debugger was disregarding my breakpoints for some reason - this is why I thought my event handler was being interrupted or removed. I resolved that by using the iPhone simulator on my Mac instead of the physical iPod touch device I was using previously to debug.

**Comment by aghassemi** [comment]
@shawnbuso Awesome, thanks for powering through quirks. Behaviour of binary attributes are certainly not consistent across browsers. I ran into a similar issue with `muted`.

@spacedino can help with design of the fullscreen button. Added `Needs UX` to the issue.

I might be worth bringing up the IMA requirements for fullscreen functionality with the Web Platform folks (/cc @mounirlamouri) so it can be considered for in the spec.

**Comment by mounirlamouri** [comment]
Glad to hear that you want to use native controls :) /CC @beaufortfrancois @avayvod

1. I guess nothing on the spec side here.
2. Would you be interested to have a way to request controls without a seek bar?
3. We launched recently an API to disable a few buttons on the HTMLMediaElement, see https://developers.google.com/web/updates/2017/03/chrome-58-media-updates#controlslist Among other things, you can hide the `fullscreen` button. If you want to target older versions of Chrome, you can use the CSS selector solution but I wouldn't recommend this solution in the long term.

**Comment by aghassemi** [comment]
@mounirlamouri `controlslist` is great! Didn't know about it. 
RE: fullscreen, I believe what would help here is ability to change the `onclick` behaviour of the native fullscreen button (My understanding is that IMA likes to take a totally different element to fullscreen rather than the video itself @shawnbuso correct me if I am wrong). Current workaround is to hide the native fullscreen button and shown our own but that is not great on the UX side as the custom fullscreen button will look different/ be in different location than the native.

**Comment by shawnbuso** [comment]
@mounirlamouri @aghassemi Correct, changing the behavior of the fullscreen button would be most ideal. I currently have my own button that works as needed, and I was able to hide the fullscreen button from the native controls by removing 'allowfullscreen' from the parent iframe attributes. The thing I'm wrestling with now is syncing showing/hiding my fullscreen button with the native controls - it looks like there's no way for me to detect when the native controls are shown or hidden, so I'm not sure how to best sync the two. If I could just re-purpose the native controls' fullscreen button I could avoid this issue (and all the UX work too, which would be nice :) )

**Comment by shawnbuso** [comment]
Any additional insight on this one? Like I mentioned earlier I've got this more or less working, I just need to figure out how to best sync showing and hiding the fullscreen button with the rest of the player controls. Are there some hard and fast rules about when the video player will show and hide controls that I can mimic in my own code? It seems to be browser and platform dependent:

on desktop the controls show when I mouse over and hide immediately on mouse out
on Android Chrome the controls show on click and hide after about 3 seconds
on iPhone and iPad they show on click but hide after about 4 seconds

I could write a decision tree that changes my show/hide behavior based on the platform and browser, but it would be much cleaner if I could just ask the player to tell me when it shows and hides controls :/

**Comment by aghassemi** [comment]
@shawnbuso for now I think we should special case the popular platforms: Android-Chrome, iOS-* (iOS should be the same regardless of browser) and for everything else do a reasonable default (show/hide on tap/hover and do 3 second timeout)

**Comment by shawnbuso** [comment]
Can we re-assign this? Thanks :)

**Comment by aghassemi** [comment]
Closing in favor of https://github.com/ampproject/amphtml/issues/10502

**Comment by aghassemi** [comment]
Reopening this. With both iOS and Chrome updating their UI controls, I see much less reasons to support custom designed controls in AMP. 

Would be ideal if we can switch IMA to use native controls. Current blockers to do so are listed in the description of the issue.

**Comment by curseagain** [comment]
@aghassemi, what is the status on this?  We spoke earlier and it would be great to just make a quick fix to add some missing controls in amp-ima-video.

There are some other buggy things about amp-ima-video that may or may not need to have issues assigned.

- missing controls (mute/unmute, replay)
- unable to loop - and unable to determine when video or ad has ended so that we can trigger replay
- when you click on video or ad the sound comes on and there is no mute button
- when in fullscreen and ad is playing there is no way to exit the ad or mute

You should be able to replicate all of these issues in amp-by-example:
![amp-ima-video](https://user-images.githubusercontent.com/4807680/46101990-1cecf280-c182-11e8-87fe-89958d1a70dd.png)

cc @torch2424

**Comment by torch2424** [comment]
This was brought up during AMP Contributor Summit 😄 

@curseagain is interested in splitting this up to smaller tasks that we can split amongst our teams to ease the bandwidth.

@aghassemi Let us know if this sounds good.

**Comment by alanorozco** [comment]
@curseagain @torch2424 

Our current plans is to support IMA via a third-party video plugin, and be able to use something like Video.JS. That way we ease our maintenance load and we also provide a much better player at the same time.

Maybe adding mute/unmute would be fine, since that seems simple enough for a temporary fix. Be aware that the approach will change to use Video.JS (which will include mute/unmute as well).

**Comment by torch2424** [comment]
So I've created three issues to Granularlize new requests from @curseagain:

#18531
#18532
#18533

Let's continue discussion there, and figure out what we can do fix small bugs on the existing `amp-ima-video`.

@aghassemi @alanorozco 

Is there any issue open for the rebuild? I feel like it would be good to close this, and just start referring new discussion of `<amp-ima-video>` there, since we now have this dialed down.

**Comment by aghassemi** [comment]
Rebuilt of `amp-ima-video` is a not sure thing yet, really comes down to priority. With `amp-video-iframe` launching soon, now there will be at least a decent work-around. I will leave it with @jasti to prioritize support and/or rebuilt of AMP IMA video. Either by us or the IMA team.

**Comment by jasti** [comment]
It was decided offline that the AMP team will continue to add support to the `amp-ima-video` player instead of a complete rewrite.

**Comment by torch2424** [comment]
@aghassemi @jasti Is it alright if I close this? 

We've made a bunch of improvements to ima-video, and we've determined we can't use native controls. But implemented all the needs that were asked by own custom player.

**Comment by aghassemi** [comment]
Let's lower the priority but keep open.

**Comment by aghassemi** [comment]
This would be really nice. Aside from bugs with the custom controls, features such as captions are missing as well. Native is the way to go.
